File: packages/testing/src/adapters/generic.ts
/**
 * Generic Adapter
 *
 * Fallback adapter for apps that don't have a specific adapter.
 * Provides basic functionality using convention over configuration.
 */

import type {
	AdapterContext,
	AppAdapter,
	ApplyFixtureResult,
	FixtureValidationResult,
} from './types.js';

// ============================================================================
// Generic Adapter
// ============================================================================

/**
 * Create a generic adapter for an app.
 * Uses conventions to provide basic functionality.
 */
export function createGenericAdapter(
	name: string,
	options: GenericAdapterOptions = {}
): AppAdapter {
	const {
		version = '1.0.0',
		cliCommand = name,
		defaultArgs = [],
		inputFormat = 'json-arg',
		outputFormat = 'json',
		commands = [],
		errors = [],
		jobs = [],
	} = options;

	return {
		name,
		version,

		cli: {
			command: cliCommand,
			defaultArgs,
			inputFormat,
			outputFormat,
		},

		fixture: {
			async apply(fixture: unknown, context: AdapterContext): Promise<ApplyFixtureResult> {
				return genericFixtureApplicator(fixture, context);
			},

			async reset(_context: AdapterContext): Promise<void> {
				// Generic reset - no-op by default
				// Apps should override this if they need cleanup
			},

			async validate(fixture: unknown): Promise<FixtureValidationResult> {
				return genericFixtureValidator(fixture);
			},
		},

		commands: {
			list: () => commands,
			getDescription: (cmd: string) => `Execute ${cmd} command`,
		},

		errors: {
			list: () => errors,
			getDescription: (code: string) => `Error: ${code}`,
			isRetryable: (code: string) => ['TIMEOUT', 'NETWORK_ERROR', 'RATE_LIMITED'].includes(code),
		},

		jobs: {
			list: () => jobs,
			getDescription: (job: string) => `Job: ${job}`,
		},
	};
}

/**
 * Options for creating a generic adapter.
 */
export interface GenericAdapterOptions {
	/** App version */
	version?: string;

	/** CLI command to run */
	cliCommand?: string;

	/** Default CLI arguments */
	defaultArgs?: string[];

	/** How to pass JSON input */
	inputFormat?: 'json-arg' | 'json-stdin' | 'flags';

	/** Expected output format */
	outputFormat?: 'json' | 'text';

	/** List of known commands */
	commands?: string[];

	/** List of known error codes */
	errors?: string[];

	/** List of known jobs */
	jobs?: string[];
}

// ============================================================================
// Generic Fixture Handling
// ============================================================================

/**
 * Generic fixture applicator that handles common patterns.
 */
async function genericFixtureApplicator(
	fixture: unknown,
	context: AdapterContext
): Promise<ApplyFixtureResult> {
	const appliedCommands: ApplyFixtureResult['appliedCommands'] = [];
	const warnings: string[] = [];

	if (!isGenericFixture(fixture)) {
		warnings.push('Fixture does not match expected generic format');
		return { appliedCommands, warnings };
	}

	const { handler } = context;
	if (!handler) {
		warnings.push('No command handler provided, fixture not applied');
		return { appliedCommands, warnings };
	}

	// Handle 'data' array - generic data seeding
	if (fixture.data && Array.isArray(fixture.data)) {
		for (const item of fixture.data) {
			if (isDataItem(item)) {
				try {
					const result = await handler(item.command, item.input ?? {});
					appliedCommands.push({
						command: item.command,
						input: item.input,
						result,
					});
				} catch (error) {
					warnings.push(
						`Failed to apply ${item.command}: ${error instanceof Error ? error.message : String(error)}`
					);
				}
			}
		}
	}

	// Handle 'setup' commands - explicit setup steps
	if (fixture.setup && Array.isArray(fixture.setup)) {
		for (const step of fixture.setup) {
			if (isDataItem(step)) {
				try {
					const result = await handler(step.command, step.input ?? {});
					appliedCommands.push({
						command: step.command,
						input: step.input,
						result,
					});
				} catch (error) {
					warnings.push(
						`Failed setup step ${step.command}: ${error instanceof Error ? error.message : String(error)}`
					);
				}
			}
		}
	}

	return { appliedCommands, warnings };
}

/**
 * Generic fixture validator.
 */
async function genericFixtureValidator(fixture: unknown): Promise<FixtureValidationResult> {
	const errors: string[] = [];

	if (typeof fixture !== 'object' || fixture === null) {
		errors.push('Fixture must be an object');
		return { valid: false, errors };
	}

	const obj = fixture as Record<string, unknown>;

	// Check for app identifier
	if (!obj.app || typeof obj.app !== 'string') {
		errors.push("Fixture should have an 'app' field identifying the target app");
	}

	// Validate data array if present
	if (obj.data !== undefined) {
		if (!Array.isArray(obj.data)) {
			errors.push("Fixture 'data' must be an array");
		} else {
			for (let i = 0; i < obj.data.length; i++) {
				const item = obj.data[i];
				if (!isDataItem(item)) {
					errors.push(`data[${i}] must have a 'command' field`);
				}
			}
		}
	}

	// Validate setup array if present
	if (obj.setup !== undefined) {
		if (!Array.isArray(obj.setup)) {
			errors.push("Fixture 'setup' must be an array");
		} else {
			for (let i = 0; i < obj.setup.length; i++) {
				const step = obj.setup[i];
				if (!isDataItem(step)) {
					errors.push(`setup[${i}] must have a 'command' field`);
				}
			}
		}
	}

	return {
		valid: errors.length === 0,
		errors: errors.length > 0 ? errors : undefined,
	};
}

// ============================================================================
// Type Guards
// ============================================================================

interface GenericFixture {
	app?: string;
	data?: unknown[];
	setup?: unknown[];
	[key: string]: unknown;
}

interface DataItem {
	command: string;
	input?: Record<string, unknown>;
}

function isGenericFixture(value: unknown): value is GenericFixture {
	return typeof value === 'object' && value !== null;
}

function isDataItem(value: unknown): value is DataItem {
	return (
		typeof value === 'object' &&
		value !== null &&
		'command' in value &&
		typeof (value as Record<string, unknown>).command === 'string'
	);
}

// ============================================================================
// Default Generic Adapter
// ============================================================================

/**
 * Default generic adapter instance.
 * Used as fallback when no specific adapter is found.
 */
export const genericAdapter = createGenericAdapter('generic', {
	version: '1.0.0',
	cliCommand: 'app',
});

File: python/src/afd/testing/adapters/generic.py
"""
Generic adapter implementation.

Fallback adapter for apps that don't have a specific adapter.
Provides basic functionality using convention over configuration.

Port of packages/testing/src/adapters/generic.ts
"""

from __future__ import annotations

from typing import Any

from afd.testing.adapters.types import (
	AdapterCliConfig,
	AdapterContext,
	AppAdapter,
	AppliedCommand,
	ApplyFixtureResult,
	CommandsConfig,
	ErrorsConfig,
	FixtureConfig,
	FixtureValidationResult,
	JobsConfig,
)


def create_generic_adapter(
	name: str = "generic",
	version: str = "1.0.0",
	*,
	cli_command: str | None = None,
	default_args: list[str] | None = None,
	input_format: str = "json-arg",
	output_format: str = "json",
	commands: list[str] | None = None,
	errors: list[str] | None = None,
	jobs: list[str] | None = None,
) -> AppAdapter:
	"""Create a generic adapter for an app."""
	cmd_list = commands or []
	err_list = errors or []
	job_list = jobs or []

	async def apply_fixture(fixture: Any, context: AdapterContext) -> ApplyFixtureResult:
		return await _generic_fixture_applicator(fixture, context)

	async def reset_fixture(_context: AdapterContext) -> None:
		pass

	async def validate_fixture(fixture: Any) -> FixtureValidationResult:
		return await _generic_fixture_validator(fixture)

	return AppAdapter(
		name=name,
		version=version,
		cli=AdapterCliConfig(
			command=cli_command or name,
			default_args=default_args,
			input_format=input_format,
			output_format=output_format,
		),
		fixture=FixtureConfig(
			apply=apply_fixture,
			reset=reset_fixture,
			validate=validate_fixture,
		),
		commands=CommandsConfig(
			list=lambda: cmd_list,
			get_description=lambda cmd: f"Execute {cmd} command",
		),
		errors=ErrorsConfig(
			list=lambda: err_list,
			get_description=lambda code: f"Error: {code}",
			is_retryable=lambda code: code in {"TIMEOUT", "NETWORK_ERROR", "RATE_LIMITED"},
		),
		jobs=JobsConfig(
			list=lambda: job_list,
			get_description=lambda job: f"Job: {job}",
		),
	)


async def _generic_fixture_applicator(
	fixture: Any, context: AdapterContext
) -> ApplyFixtureResult:
	"""Generic fixture applicator that handles common patterns."""
	applied_commands: list[AppliedCommand] = []
	warnings: list[str] = []

	if not _is_generic_fixture(fixture):
		warnings.append("Fixture does not match expected generic format")
		return ApplyFixtureResult(applied_commands=applied_commands, warnings=warnings)

	handler = context.handler
	if not handler:
		warnings.append("No command handler provided, fixture not applied")
		return ApplyFixtureResult(applied_commands=applied_commands, warnings=warnings)

	for array_key in ("data", "setup"):
		items = fixture.get(array_key)
		if items and isinstance(items, list):
			for item in items:
				if _is_data_item(item):
					try:
						result = await handler(item["command"], item.get("input", {}))
						applied_commands.append(
							AppliedCommand(
								command=item["command"],
								input=item.get("input"),
								result=result,
							)
						)
					except Exception as err:
						warnings.append(
							f"Failed to apply {item['command']}: {err}"
						)

	return ApplyFixtureResult(applied_commands=applied_commands, warnings=warnings)


async def _generic_fixture_validator(fixture: Any) -> FixtureValidationResult:
	"""Generic fixture validator."""
	errors: list[str] = []

	if not isinstance(fixture, dict):
		errors.append("Fixture must be an object")
		return FixtureValidationResult(valid=False, errors=errors)

	if not fixture.get("app") or not isinstance(fixture.get("app"), str):
		errors.append("Fixture should have an 'app' field identifying the target app")

	for array_key in ("data", "setup"):
		if array_key in fixture:
			arr = fixture[array_key]
			if not isinstance(arr, list):
				errors.append(f"Fixture '{array_key}' must be an array")
			else:
				for i, item in enumerate(arr):
					if not _is_data_item(item):
						errors.append(f"{array_key}[{i}] must have a 'command' field")

	return FixtureValidationResult(
		valid=len(errors) == 0,
		errors=errors if errors else None,
	)


def _is_generic_fixture(value: Any) -> bool:
	return isinstance(value, dict)


def _is_data_item(value: Any) -> bool:
	return isinstance(value, dict) and isinstance(value.get("command"), str)
